### ISSUE TITLE
RFC: publish a compose application on registry

### ISSUE THREAD

**Issue by ndeloof** [compose-spec/compose-spec#288]
**What is the problem you're trying to solve**
On can easily push all images involves in a Compose application, but still need to find another way to share the compose.yaml definition.

I suggest we define a mechanism and media type to push a compose.yaml file to a registry, in addition to images, with adequate OCI references

**Describe the solution you'd like**
Push Compose application as an [OCI artifact](https://github.com/opencontainers/image-spec/blob/main/artifact.md).
We would use:
- `application/vnd.docker.compose.application` media type for the whole compose application
- `application/vnd.docker.compose.config+yaml` media type for the compose.yaml model
- OCI or Docker manifest media type for referenced images

```json
{
   "schemaVersion": 2,
  "mediaType": "application/vnd.oci.artifact.manifest.v1+json",
  "artifactType": "application/vnd.docker.compose.application",
   "blobs": [
      {
		"mediaType": "application/vnd.docker.compose.config+yaml",
		"digest": "sha256:.."
		"size": 123,
	  },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "digest": "sha256:.",
         "size": 123,
      },
```

Mechanism to push/pull such a manifest is our of the scope of the compose-specification, but for illustration this could result into something like this:
```
$ docker compose push --application registry/foo/bar
[+] Running 3/3
 ⠼ Pushing foo: xxxx Layer already exists                                                                         2.4s
 ⠸ Pushing compose.yaml: xxxx                                                                                         2.4s
 ⠸ Pushing compose application: xxxx                                                                              2.4s

$ docker compose --application registry/foo/bar  up 
...
```

**Additional context**
This somehow resurrects the main concept of [Docker App](https://github.com/docker/app)

**Comment by EricHripko** [comment]
I think sharing Compose files is a very valuable feature to have - particularly, in a SOA/non-monorepo scenario. Providing every service lives in a different repository, there isn't a good place to store the Compose model that "glues" these together. Hosting these in the OCI registry next to images could be a great way to fill this gap 🙂

**Comment by asim** [comment]
No updates as of late, is there anything stopping this from moving forward?

**Comment by ndeloof** [comment]
@asim no "blocker", just open discussion on the adequate approach to adopt, then need to be implemented by tooling

**Comment by asim** [comment]
> 

👍 Look forward to seeing it in action!

**Comment by ndeloof** [comment]
Ran a quick experiment, found a few issues with possible workarounds:

1. service image references

Use of a manifest-list type as suggested on https://github.com/compose-spec/compose-spec/issues/288#issuecomment-1344065157 allow to get support by existing registries but comes with a limitation: sub-manifests referred by digest need to be stored in the same repository. So if I publish my compose application under `ndeloof/foo`, my compose.yaml config artifact need to be pushed under same repository and referred by digest (fine) but so do the images used in my app. I can't just refer to `nginx@sha256:1234`, I need this image pushed into my own repository

Docker engine doesn't allow to push without a tag, so I would have to tag all compose images as `ndeloof/foo:sometag`. This would polute my repository with many irrelevant tags. 
As an alternative, we can use registryClient to [mount](https://docs.docker.com/registry/spec/api/#cross-repository-blob-mount) service images into repository, without a tag and refer to those by digest.

Then we eventually need to bake the top-level manifest list and push it to registry.

I pushed a draft (non-functional) implementation of this on https://github.com/docker/compose/tree/multiplayer relying on `oras` CLI for this last step, would need to use equivalent feature implemented by [oras-go](https://github.com/oras-project/oras-go) for full integration but I haven't found an obvious way to set registry credentials 😅

2. published compose file

The compose application model being parsed to run such a `publish` command is a parsed/interpolated flavor of the original compose.yaml. If we marshall it into yaml this will just reflect user's environment and project name and make this hardly reusable by others. So we need to publish the original compose.yaml file (which might also include some useful documentation as yaml comments). Challenge then is that service images are not resolved by digest in this file, so we would publish a compose.yaml with service referring to `nginx`, which is not the same as the `nginx@sha256:1234` we resolved on (1) and mounted as `docker/foo@sha256:1234`.  So we miss a way to replicate the yaml merge and image replace logic but preserve the yaml file as much as possible.

A possible (but ugly?) way would be to push the original compose.yaml file(s) AND append a generated one that would only declare all services with an override on service.image, so that after all yaml files get merged the resulting model would actually use `docker/foo@sha256:1234` as expected.

Another option would be to rely on a pure yaml processing library like [yq](https://github.com/mikefarah/yq) to make such a replacement in the raw yaml file(s).


so, basically, still some work needed :)
